### PR TITLE
Coreclr compatibility: Type -> TypeInfo 

### PR DESCRIPTION
**Pull request by pherbel**
I created a new PR for the TypeInfo changes.

Related to: Port Orleans to coreclr #368.
It was the first PR: #590

### CODE REVIEW

File: src/Orleans/AssemblyLoader/AssemblyLoader.cs [hunk 1]
```diff
@@ -106,8 +106,38 @@ internal class AssemblyLoader
         {
             try
             {
-                var assembly = Assembly.Load(assemblyName);
-                var foundType = TypeUtils.GetTypes(assembly, type => typeof (T).IsAssignableFrom(type)).First();
+                var exeRoot = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
```
Comment (gabikliot) [inline src/Orleans/AssemblyLoader/AssemblyLoader.cs]: Lets leave that code as is for now. we had a lot of issues with the right way to load assemblies, and Sergey has been changing that recently too. Lets deal with assembly loading later on, even if it means leaving the type usage as is here for now, ok?

File: src/Orleans/CodeGeneration/GrainFactoryBase.cs [hunk 2]
```diff
@@ -152,9 +153,9 @@ public static class GrainFactoryBase
             {
                 throw new ArgumentException("Cannot fabricate grain-reference for non-grain type: " + interfaceType.FullName);
             }
-            var implementation = TypeCodeMapper.GetImplementation(interfaceType, grainClassNamePrefix);
+            var implementation = TypeCodeMapper.GetImplementation(interfaceType.GetTypeInfo(), grainClassNamePrefix);
```
Comment (gabikliot) [inline src/Orleans/CodeGeneration/GrainFactoryBase.cs]: can call GetTypeInfo once and resuse.

File: src/Orleans/CodeGeneration/GrainInterfaceData.cs [hunk 3]
```diff
@@ -58,43 +58,43 @@ public RulesViolationException(string message, List<string> violations)
 
         public bool IsExtension
         {
-            get { return typeof(IGrainExtension).IsAssignableFrom(Type); }
+            get { return typeof(IGrainExtension).GetTypeInfo().IsAssignableFrom(TypeInfo); }
```
Comment (gabikliot) [inline src/Orleans/CodeGeneration/GrainInterfaceData.cs]: can cache `typeof(IGrainExtension).GetTypeInfo()`  in a static readonly

File: src/Orleans/CodeGeneration/GrainInterfaceData.cs [hunk 4]
```diff
@@ -104,49 +104,49 @@ public GrainInterfaceData(Language language)
...
             return gi;
         }
 
-        public static bool IsGrainInterface(Type t)
+        public static bool IsGrainInterface(TypeInfo t)
         {
             if (t.IsClass)
                 return false;
-            if (t == typeof(IGrainObserver) || t == typeof(IAddressable) || t == typeof(IGrainExtension))
+            if (t.Equals(typeof(IGrainObserver)) || t.Equals(typeof(IAddressable)) || t.Equals(typeof(IGrainExtension)))
```
Comment (gabikliot) [inline src/Orleans/CodeGeneration/GrainInterfaceData.cs]: does that work correctly? typeInfo equals type?

File: src/Orleans/Serialization/SerializationManager.cs [hunk 5]
```diff
@@ -921,11 +925,11 @@ public static void SerializeInner(object obj, BinaryTokenStreamWriter stream, Ty
                 return;
             }
 
-            var t = obj.GetType();
+            var typeInfo = obj.GetType().GetTypeInfo();
```
Comment (gabikliot) [inline src/Orleans/Serialization/SerializationManager.cs]: This place scares me most! It is very hot path, called for every argument in every grain call. Do we need both GetType().GetTypeInfo();?

File: src/OrleansRuntime/Catalog/Catalog.cs [hunk 6]
```diff
@@ -277,7 +278,7 @@ public void UnregisterMessageTarget(ActivationData activation)
 
             if (activation.GrainInstance == null) return;
 
-            var grainTypeName = TypeUtils.GetFullName(activation.GrainInstanceType);
+            var grainTypeName = TypeUtils.GetFullName(activation.GrainInstanceType.GetTypeInfo());
```
Comment (gabikliot) [inline src/OrleansRuntime/Catalog/Catalog.cs]: worth caching TypeInfo inside ActivatioData (init it when activationData.GrainInstanceType is inited) and reuse.

File: src/Orleans/CodeGeneration/GrainInterfaceData.cs (hunk 4 above)
Comment (pherbel) [inline src/Orleans/CodeGeneration/GrainInterfaceData.cs]: Yes. I've checked. Actually, the TypeInfo does not have '==' operator in coreclr so you should use the TypeInfo.Equals method.

File: src/Orleans/CodeGeneration/GrainInterfaceData.cs (hunk 3 above)
Comment (pherbel) [inline src/Orleans/CodeGeneration/GrainInterfaceData.cs]: Sure!

File: src/Orleans/AssemblyLoader/AssemblyLoader.cs (hunk 1 above)
Comment (pherbel) [inline src/Orleans/AssemblyLoader/AssemblyLoader.cs]: I will revert the AssemblyLoader changes

File: src/Orleans/CodeGeneration/GrainFactoryBase.cs (hunk 2 above)
Comment (pherbel) [inline src/Orleans/CodeGeneration/GrainFactoryBase.cs]: Sure!

File: src/Orleans/Serialization/SerializationManager.cs (hunk 5 above)
Comment (pherbel) [inline src/Orleans/Serialization/SerializationManager.cs]: I will check it how we can do it.

File: src/OrleansRuntime/Catalog/Catalog.cs (hunk 6 above)
Comment (pherbel) [inline src/OrleansRuntime/Catalog/Catalog.cs]: Sure!

**Comment by gabikliot** [comment]
Here is a list of cold vs. hot code paths:
1) All ClientGenerator, including ClientGenerator/Serilalization is NOT under hot path. Can stay as is.
2) Lets keep the whole Assembly Loader code as is for now and deal with it later on.
3) GrainFactoryBase, GrainFactory, TypeCodeMappper are called every time we do GetGrain, so not super-hot path, but still optimizing for trivial variable reuse, if there is an opportunity (if there are multiple calls for the same GetTypeInfo in the same func use local var to cache it).
4) GrainInterfaceData and TypeUtils – a lot of utility functions that are used everywhere. A bit hard to tell for sure about hot vs. cold path. Mostly on the cold path, but still worth easy optimizations, if possible, by local re-use where possible. 
5) GlobalConfiguration, ProviderLoader and ProviderTypeLoader are cold path.
6) GrainInterfaceMap.AddEntry is cold path
7) OutsideGrainClient.Load provider is cold path
8) Serialization – ufffff – that is a hard one. In general VERY Hot path, some places may be called multiple times for every msg. Worth considering adding a concurrent dictionary from type to TypeInfo just for serialization code. This whole code, especially the SerializatonManager is the place to watch and will be a primary suspect for degraded perf.
9) ImplicitSubsriber is cold path
10) Utils. ObjectToList InvalidCastException is cold path
11) GrainTypeData is cold path
12) SiloAssemblyLoader is cold path

**Comment by gabikliot** [comment]
@pherbel , did you get a chance to look at those comments I posted? Would you have time to make those fixes to optimize hot pass code? Thanks!

**Comment by pherbel** [comment]
@gabikliot Sorry I was on sick leave and I had vacation last weeks. I fix those things as soon as I can.

**Comment by sergeybykov** [comment]
Looks like this PR got subsumed by https://github.com/dotnet/orleans/pull/973.

**Pull request closed without merging**